File: ruby/udemy_class/recursion/string_permutations.rb
require 'byebug'

def permutations(string)
  if string.length == 1
    return [string]
  end

  string.split("").each_with_index do |char, index|
    # We want to get the remaining characters besides the current one
    characters_before = string[0...index] || ""
    characters_after = string[index + 1..-1] || ""

    characters_remaining = characters_before + characters_after

    # We want to find the permutations of all remaining characters
    # For example, if out string is "abc" and we are looking at "b"
    # We want to return a list of ["ac", "ca"]

    permutations_of_remainder = permutations(characters_remaining)

    # Once we have the permutations of the remaining characters, we add our
    # current char to each.
    # Ex perms = ["ac", "ca"]
    # After this step = ["bac", "cab"]
    permutations_of_remainder.each do |perm|
      output << char + perm
    end
  end

  return output
end

print(permutations("abc"))

File: python/udemy_class/recursion/string_permutations.py
def string_permutations(string):
    # Base case. We want to return the string if the length is one
    if len(string) == 1:
        return string

    output = []

    # Iterate through each character.
    # For each character, we want to then get the permutations of
    # all other characters in the string
    for i, character in enumerate(string):
        # find permutations of remaining characters
        permutations = string_permutations(string[:i] + string[i + 1:])

        # Once we have the permutations of all the other characters,
        # we can add the current character to the beginning of them
        # in order to make permutations inclduing the current character
        for perm in permutations:
            output.append(character + perm)

    return output


print(string_permutations("abc"))
print(string_permutations("ab"))
